### PR TITLE
feat: Multitenancy dashboard

### PR DESCRIPTION
**Pull request by sattvikc**
## Summary of change

(A few sentences about this PR)

## Related issues

-   Link to issue1 here
-   Link to issue1 here

## Test Plan

(Write your test plan here. If you changed any code, please provide us with clear instructions on how you verified your changes work. Bonus points for screenshots and videos!)

## Documentation changes

(If relevant, please create a PR in our [docs repo](https://github.com/supertokens/docs), or create a checklist here highlighting the necessary changes)

## Checklist for important updates

-   [ ] Changelog has been updated
-   [ ] `coreDriverInterfaceSupported.json` file has been updated (if needed)
    -   Along with the associated array in `lib/ts/version.ts`
-   [ ] `frontendDriverInterfaceSupported.json` file has been updated (if needed)
-   [ ] Changes to the version if needed
    -   In `package.json`
    -   In `package-lock.json`
    -   In `lib/ts/version.ts`
-   [ ] Had run `npm run build-pretty`
-   [ ] Had installed and ran the pre-commit hook
-   [ ] If new thirdparty provider is added,
    -   [ ] update switch statement in `recipe/thirdparty/providers/configUtils.ts` file, `createProvider` function.
    -   [ ] add an icon on the user management dashboard.
-   [ ] Issue this PR against the latest non released version branch.
    -   To know which one it is, run find the latest released tag (`git tag`) in the format `vX.Y.Z`, and then find the latest branch (`git branch --all`) whose `X.Y` is greater than the latest released tag.
    -   If no such branch exists, then create one from the latest released branch.
-   [ ] If have added a new web framework, update the `add-ts-no-check.js` file to include that
-   [ ] If added a new recipe / api interface, then make sure that the implementation of it uses NON arrow functions only (like `someFunc: function () {..}`).
-   [ ] If added a new recipe, then make sure to expose it inside the recipe folder present in the root of this repo. We also need to expose its types.

## Remaining TODOs for this PR

-   [ ] Item1
-   [ ] Item2

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -20,19 +20,38 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
...
+    -   GET `/api/thirdparty/config`
+    -   PUT `/api/thirdparty/config`
+    -   DELETE `/api/thirdparty/config`
 -   `passwordResetPOST`:
     -   now verifies the email address in all cases if the EmailVerification recipe is initialized
     -   now tries to link accounts based on account info if AccountLinking is enabled
 -   Extracted some tests into a separate [backend-sdk-testing](https://github.com/supertokens/backend-sdk-testing/) repo, to reuse tests between our backend SDKs
+-   Sends `websiteDomain` and `apiDomain` to core for telemetry.
+-   `boxyURL` is no more mandatory input in `additionalConfig` while adding boxy-saml provider in thirdParty.
+-   Fixes issue with OIDC discover when the input url already contains `.well-known/openid-configuration`.
```
Comment (rishabhpoddar) [inline CHANGELOG.md:54]: Need a migration section for all the breaking changes!

Comment (rishabhpoddar) [inline CHANGELOG.md:54]: other backend sdks will also need these two changes, right?

Why have we changed the boxy hq thing to not need the additional config?

File: lib/ts/QuerierError.ts [hunk 2]
```diff
@@ -0,0 +1,15 @@
+export class QuerierError extends Error {
```
Comment (rishabhpoddar) [inline lib/ts/QuerierError.ts:1]: why do we need this?

File: lib/ts/querier.ts [hunk 3]
```diff
@@ -48,11 +50,26 @@ export class Querier {
...
+
+        const queryParamsObj: {
+            apiDomain: string;
+            websiteDomain?: string;
+        } = {
+            apiDomain: appInfo.apiDomain.getAsStringDangerous(),
+        };
+        const request = st.getRequestFromUserContext(userContext);
+        if (request !== undefined) {
+            queryParamsObj.websiteDomain = appInfo.getOrigin({ request, userContext }).getAsStringDangerous();
```
Comment (rishabhpoddar) [inline lib/ts/querier.ts:69]: call this even if request is undefined

File: lib/ts/querier.ts [hunk 4]
```diff
@@ -63,7 +80,22 @@ export class Querier {
...
+                            url: url,
+                            method: "get",
+                            headers: headers,
+                            params: queryParamsObj,
+                        },
+                        userContext
+                    );
+                    url = request.url;
+                    headers = request.headers;
+                }
```
Comment (rishabhpoddar) [inline lib/ts/querier.ts:96]: this change also needs to be added to other backend sdks

File: test/dashboard/createOrUpdateThirdPartyConfig.test.js [hunk 5]
```diff
@@ -0,0 +1,111 @@
+/* Copyright (c) 2024, VRAI Labs and/or its affiliates. All rights reserved.
```
Comment (rishabhpoddar) [inline test/dashboard/createOrUpdateThirdPartyConfig.test.js:1]: all of this goes in backend-sdk-testing

File: lib/ts/recipe/dashboard/api/multitenancy/getThirdPartyConfig.ts [hunk 6]
```diff
@@ -0,0 +1,254 @@
...
+    if (providersFromCore === undefined) {
+        providersFromCore = [];
+    }
+
+    let found = false;
+    for (let i = 0; i < providersFromCore.length; i++) {
+        if (providersFromCore[i].thirdPartyId === thirdPartyId) {
+            found = true;
+            if (thirdPartyId === "okta") {
+                providersFromCore[i].oidcDiscoveryEndpoint = options.req.getKeyValueFromQuery("oktaDomain");
```
Comment (rishabhpoddar) [inline lib/ts/recipe/dashboard/api/multitenancy/getThirdPartyConfig.ts:71]: add comments here explaining what is happening

File: lib/ts/recipe/multifactorauth/recipe.ts [hunk 7]
```diff
@@ -69,8 +71,16 @@ export default class Recipe extends RecipeModule {
...
+            let originalImpl = RecipeImplementation(this);
+            let builder = new OverrideableBuilder(originalImpl);
             this.recipeInterfaceImpl = builder.override(this.config.override.functions).build();
+
+            // @ts-ignore
+            let layers = builder.layers;
+
+            if (Object.keys(layers[1]).includes("getMFARequirementsForAuth")) {
+                this.isGetMfaRequirementsForAuthOverridden = true;
+            }
```
Comment (rishabhpoddar) [inline lib/ts/recipe/multifactorauth/recipe.ts:83]: just check that the override function is provided and nothing else is needed

File: lib/ts/recipe/multitenancy/types.ts [hunk 8]
```diff
@@ -176,13 +166,12 @@ export type APIInterface = {
               emailPassword: {
                   enabled: boolean;
               };
-              passwordless: {
-                  enabled: boolean;
-              };
               thirdParty: {
-                  enabled: boolean;
```
Comment (rishabhpoddar) [inline lib/ts/recipe/multitenancy/types.ts:183]: this is wrong

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: lib/ts/recipe/dashboard/api/multitenancy/createOrUpdateThirdPartyConfig.ts [hunk 9]
```diff
@@ -0,0 +1,68 @@
+/* Copyright (c) 2024, VRAI Labs and/or its affiliates. All rights reserved.
+ *
```
Comment (rishabhpoddar) [inline lib/ts/recipe/dashboard/api/multitenancy/createOrUpdateThirdPartyConfig.ts:2]: need to make sure there are no calls to the core in a loop in any of the dashboard apis

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: lib/ts/QuerierError.ts (hunk 2 above)
Comment (sattvikc) [reply lib/ts/QuerierError.ts:1]: Removed

File: lib/ts/querier.ts (hunk 3 above)
Comment (sattvikc) [reply lib/ts/querier.ts]: Done

File: lib/ts/recipe/dashboard/api/multitenancy/getThirdPartyConfig.ts (hunk 6 above)
Comment (sattvikc) [reply lib/ts/recipe/dashboard/api/multitenancy/getThirdPartyConfig.ts:71]: Done

File: lib/ts/recipe/multifactorauth/recipe.ts (hunk 7 above)
Comment (sattvikc) [reply lib/ts/recipe/multifactorauth/recipe.ts:83]: Done

File: lib/ts/recipe/multitenancy/types.ts (hunk 8 above)
Comment (sattvikc) [reply lib/ts/recipe/multitenancy/types.ts:183]: Done

File: CHANGELOG.md (hunk 1 above)
Comment (sattvikc) [reply CHANGELOG.md:56]: Done

File: lib/ts/recipe/dashboard/api/multitenancy/createOrUpdateThirdPartyConfig.ts (hunk 9 above)
Comment (sattvikc) [reply lib/ts/recipe/dashboard/api/multitenancy/createOrUpdateThirdPartyConfig.ts:2]: Done

**Pull request merged**